retrofit-converters/kotlinx-serialization/src/main/java/com/jakewharton/retrofit2/converter/kotlinx/serialization/Serializer.kt

```diff
@@ -0,0 +1,46 @@
+package com.jakewharton.retrofit2.converter.kotlinx.serialization
+
+import java.lang.reflect.Type
+import kotlinx.serialization.BinaryFormat
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerialFormat
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.StringFormat
+import kotlinx.serialization.serializer
+import okhttp3.MediaType
+import okhttp3.RequestBody
+import okhttp3.ResponseBody
+
+internal sealed class Serializer {
+  abstract fun <T> fromResponseBody(loader: DeserializationStrategy<T>, body: ResponseBody): T
+  abstract fun <T> toRequestBody(contentType: MediaType, saver: SerializationStrategy<T>, value: T): RequestBody
+
+  protected abstract val format: SerialFormat
+
+  fun serializer(type: Type): KSerializer<Any> = format.serializersModule.serializer(type)
+
+  class FromString(override val format: StringFormat) : Serializer() {
+    override fun <T> fromResponseBody(loader: DeserializationStrategy<T>, body: ResponseBody): T {
+      val string = body.string()
+      return format.decodeFromString(loader, string)
+    }
+
+    override fun <T> toRequestBody(contentType: MediaType, saver: SerializationStrategy<T>, value: T): RequestBody {
+      val string = format.encodeToString(saver, value)
+      return RequestBody.create(contentType, string)
+    }
+  }
+
+  class FromBytes(override val format: BinaryFormat) : Serializer() {
+    override fun <T> fromResponseBody(loader: DeserializationStrategy<T>, body: ResponseBody): T {
+      val bytes = body.bytes()
```

Comment:
In theory you could also hook up the kotlinx.serialization Okio integration to get streaming. This is fine though.

Reply:
Streaming is only available for JSON. Not arbitrary `BinaryFormat` or `TextFormat`s.

Reply:
Ooooh that’s a thing for them to fix.

I expect JSON is like 90% of kotlinx.serialization usage. It’s great that it supports XML and protobuf and others, but I expect JSON is where specializations like this are most likely to pay off.

Definitely not for this PR.